### ISSUE TITLE
changefeedccl: test pubsub performance

### ISSUE THREAD

**Issue by jayshrivastava** [cockroachdb/cockroach#115536]
In https://github.com/cockroachdb/cockroach/issues/111829, we strongly believe that there are performance issues with the pubsub sink. We are likely emitting small batches which is bottlenecking throughput. We must complete these issues to add observability 

- [ ] https://github.com/cockroachdb/cockroach/issues/114027
- [x] https://github.com/cockroachdb/cockroach/issues/114140
- [x] https://github.com/cockroachdb/cockroach/issues/114141

Then, we need to run roachtests and observe the performance of the sink, making the necessary adjustments as needed.

Jira issue: CRDB-34088

**Comment by jayshrivastava** [comment]
TLDR: We have a performance issue in the batching sinks where if we have too many parallel emissions, processing the results as they arrive gets slower because on each successful result, we need to do a union over all the keys which are being emitted in parallel. Adding a batch to be sent and sending it is very fast, but processing the results is very slow. This bottlenecks the whole changefeed. 

More details: 

We have **nonzero** `emitted_messages` but **zero** in `sink_io_inflight` batches. Looking at this code, zero in flight messages means that no batches are in transit or being emitted. `emitted_messages` is nonzero because we have a queue of IO responses from pubsub which have been emitted successfully, but the parallel io needs to go through this queue and execute callbacks for each result. In these callbacks, we inc `emitted_messages` metric. The reason for why no messages are being emitted by the batching sink is because we've used up all the allocs and are waiting for all the callbacks to run to free up 50% of the changefeed quota. This is okay. The problem is how slow we're processing the results.

The callbacks are super slow because (a) one thread calls these callbacks and we may have hundereds of results to process and (b) for each result, we calculate the union of keys with the current set of in flight and pending keys to  

<img width="1004" alt="image" src="https://github.com/cockroachdb/cockroach/assets/18633281/1e9bca46-0eea-4894-bdd5-b9bf53ddb771">

**Comment by jayshrivastava** [comment]
The parallel IO queue nanos are extremely high. 

<img width="1127" alt="image" src="https://github.com/cockroachdb/cockroach/assets/18633281/cc0e2282-306c-47a4-b3c1-4ff85a3e110e">
